lib/spack/spack/solver/asp.py

```diff
@@ -1674,6 +1688,98 @@ def dependency_holds(input_spec, requirements):
+            # the spliceable config of the package always includes the variant
+            if vari_defs == [] or any(cond_spec.satisfies(s) for (s, _) in vari_defs):
+                variant = vari_defs[0][1]
+                if variant.multi:
+                    continue  # cannot automatically match multi-valued variants
+                value_var = AspVar(f"VariValue{i}")
+                attr_constraint = fn.attr("variant_value", splice_node, variant_name, value_var)
+                hash_attr_constraint = fn.hash_attr(
+                    hash_asp_var, "variant_value", splice_spec.name, variant_name, value_var
+                )
+                variant_constraints.append(attr_constraint)
+                variant_constraints.append(hash_attr_constraint)
+        return variant_constraints
+
+    def package_splice_rules(self, pkg):
+        self.gen.h2("Splice rules")
+        for i, (cond, (spec_to_splice, match_variants)) in enumerate(
+            sorted(pkg.splice_specs.items())
+        ):
+            with named_spec(cond, pkg.name):
+                self.version_constraints.add((cond.name, cond.versions))
+                self.version_constraints.add((spec_to_splice.name, spec_to_splice.versions))
+                when_spec_attrs = []
+                splice_spec_hash_attrs = []
+                hash_var = AspVar("Hash")
+                splice_node = fn.node(AspVar("NID"), cond.name)
+                explicit_variants = []
+                for c in self.spec_clauses(cond, body=True, required_from=None):
+                    args = c.args
+                    if args[0] == "node":
+                        continue
+                    if args[0] == "variant_value":
+                        explicit_variants.append(args[2])
+                    when_spec_attrs.append(fn.attr(args[0], splice_node, *args[2:]))
+
+                for c in self.spec_clauses(spec_to_splice, body=True, required_from=None):
+                    args = c.args
+                    if args[0] == "node":
+                        continue
+                    if args[0] == "variant_value":
+                        explicit_variants.append(args[2])
+                    splice_spec_hash_attrs.append(fn.hash_attr(hash_var, *args))
+                if match_variants is None:
+                    variant_constraints = []
+                elif match_variants == "*":
+                    filt_match_variants = set()
+                    for map in pkg.variants.values():
+                        for k in map:
+                            filt_match_variants.add(k)
+                    filt_match_variants = list(filt_match_variants)
+                    variant_constraints = self._gen_match_variant_splice_constraints(
+                        pkg, cond, spec_to_splice, hash_var, splice_node, filt_match_variants
+                    )
+                else:
+                    if any(v in explicit_variants for v in match_variants):
```

Comment:
```suggestion
                    if any(v in cond.variants or v in spec_to_splice.variants for v in match_variants):
```
To go with the change above